Enable automatic reload of dynamic job rules on change

natefoo opened the pull request:
usegalaxy.eu has a handy microservice ([usegalaxy-eu/jcaas](https://github.com/usegalaxy-eu/jcaas/)) for rapid job config changes, which is a great feature that I want, but I'd prefer not to use a web service if possible. I've instead gone the route of making job rules dynamically reloadable when their contents change.

This is configurable with the `watch_job_rules` option in `galaxy.yml` and is disabled by default so as not to change the current behavior.

This largely steals the tool config watcher code and results in some duplication, but unfortunately I don't have the bandwidth to properly refactor it right now.

natefoo committed:
Enable dynamic reload of dynamic job rules when configured, using
watchdog notifications.

hexylena commented on lib/galaxy/webapps/galaxy/config_watchers.py:
```diff
@@ -25,12 +33,22 @@ def __init__(self, app):
+                job_rules_directory,
+                callback=partial(reload_job_rules, self.app),
+                recursive=True,
+                extensions=('.py',))
```
Am I reading this correctly, does this only watch for changes in .py files? All of our rules are loaded through .yaml files in the same directory.

hexylena replied:
(And we'd rather not re-read the yaml file every time we check a job rule)

natefoo replied:
So the rule would load and cache the YAML at module load time? Yeah, I can make this configurable, or, maybe just ignore `.py[oc]`?

hexylena replied:
yeah that's how ours works now. That data structure is essentially static and tiny to hold in memory, no reason to go to disk before every job scheduling query.

Ignoring .py[oc] sounds good too!

natefoo committed:
Fix non-Galaxy webapp unit tests

natefoo committed:
Watch any objects in job rule directories that are not python bytecode,
rather than only watching python source.

natefoo commented:
Is it possible to see the details of that selenium test that failed?

martenson commented:
@natefoo https://jenkins.galaxyproject.org/job/docker-selenium-multi/346/SELENIUM_TESTS=h/testReport/junit/selenium_tests.test_history_panel_collections/HistoryPanelCollectionsTestCase/test_name_tags_display/ ?

dannon commented:
@natefoo It should be this one:  https://jenkins.galaxyproject.org/job/docker-selenium-multi/349/SELENIUM_TESTS=u-z/testReport/selenium_tests.test_workflow_run/WorkflowRunTestCase/test_simple_execution/

Looks like it was just a timeout to me?

natefoo commented:
Right, I saw that it was a timeout, just didn't know if I caused it.

I guess there is no way to see where exactly it timed out?

martenson commented:
@natefoo you can sometimes find useful screenshots in the test artifacts

jmchilton commented:
There is a stack trace on that page that indicates Galaxy was waiting for a history to become terminal that didn't become terminal in time.

```
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/galaxy/test/selenium_tests/framework.py", line 148, in func_wrapper
    return f(self, *args, **kwds)
  File "/galaxy/test/selenium_tests/framework.py", line 92, in func_wrapper
    f(self, *args, **kwds)
  File "/galaxy/test/selenium_tests/test_workflow_run.py", line 27, in test_simple_execution
    self.wait_for_history()
  File "/galaxy/test/galaxy_selenium/navigates_galaxy.py", line 234, in wait_for_history
    final_state = self.wait(timeout=timeout).until(history_becomes_terminal)
  File "/galaxy_venv/local/lib/python2.7/site-packages/selenium/webdriver/support/wait.py", line 80, in until
    raise TimeoutException(message, screen, stacktrace)
```

Under test artifacts it is available as a file download and such.

I think this timed out for the same reason the API tests randomly fail, Galaxy targeting Postgres just sometimes hangs for a minute or three. It always eventually recovers sometimes not in time for the tests. I don't know what is happening, I think there is some DB contention to blame but I don't have any proof.

natefoo committed:
Ignore change events for dot files. This could be configurable if it's
needed but I can't imagine when it would be. Prevents reloading on vim
swapfile writes.

mvdbeek commented on lib/galaxy/util/watcher.py:
```diff
@@ -0,0 +1,169 @@
+        self.observer.schedule(self.event_handler, dir, recursive=recursive)
+
+    def watch_file(self, file, callback=None):
+        file = os.path.abspath(file)
```
do we maybe want to change `file` to `path` or `fpath`, `file_path`  or some other non-builtin ? Same with `dir`

mvdbeek reviewed (approved):
LGTM, thanks @natefoo! We should indeed follow up and adapt the tool-oriented watchers!

dannon committed:
Avoid using builtins for variable names.  file->file_path, dir->dir_path in generic watcher functions

natefoo committed:
Merge pull request #8 from dannon/dynamic-rule-dynamic-reload

Avoid using builtins for variable names.
